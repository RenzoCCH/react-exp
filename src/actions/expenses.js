import uuid from 'uuid';

//ADD_EXPENSE
export const addExpense = (
  {
    id,
    description,
    note ,
    amount = 0 ,
    createdAt = 0
  } = {}
)=>({
  type:'ADD_EXPENSE',
  expense:{
    id:uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

//REMOVE_EXPENSE
export const removeExpense = ({id} = {})=>({
  type:'REMOVE_EXPENSE',
  id
});

//EDIT_EXPENSE
export const editExpense = (
  expense
)=>({
  type:'EDIT_EXPENSE',
  expense
});

