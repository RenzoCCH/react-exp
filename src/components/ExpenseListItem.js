import React from 'react';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ({id,description,amount,createdAt})=>{
  return (
    <section>
      <h3>
        <Link to={`/edit/${id}`}>{description}</Link>
      </h3>
      {amount !== 0 && <p><strong>Amount:</strong><span>{amount}</span></p>}
      {createdAt !== 0 && <p><strong>Created at:</strong><span>{createdAt}</span></p>}
    </section>
  );
};

export default ExpenseListItem;



