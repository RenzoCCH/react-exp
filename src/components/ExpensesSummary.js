import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from "../selectors/expenses";
import totalExpenses from "../selectors/expenses-total";
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <p>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}</p>
  );
};


const mapStateToProps = (state) => {
  let expenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: expenses.length,
    expensesTotal: totalExpenses(expenses)
  };
};



export default connect(mapStateToProps)(ExpensesSummary);
