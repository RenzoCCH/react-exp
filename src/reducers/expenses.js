//Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState,action = {})=>{
  switch (action.type) {
  case 'ADD_EXPENSE':
    return [...state,action.expense];
  case 'REMOVE_EXPENSE':
    return state.filter(({id})=>action.id!==id);
  case 'EDIT_EXPENSE':
    return state.map((exp)=>{
      if(exp.id === action.expense.id){
        return {...exp, ...action.expense};
      }
      else return exp;
    });
  default:
    return state;
  }
};

