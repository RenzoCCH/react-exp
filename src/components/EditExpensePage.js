import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onEditExpense = ( expense )=>{
    this.props.editExpense({...expense,id:this.props.expense.id});
    this.props.history.push('/');
  }

  onRemoveExpense =  () => {
    this.props.removeExpense({ id:this.props.expense.id });
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        {this.props.match.params.id}
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onEditExpense}
        ></ExpenseForm>
        <button onClick={this.onRemoveExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense)=>expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    editExpense  : (expense) => dispatch(editExpense(expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);