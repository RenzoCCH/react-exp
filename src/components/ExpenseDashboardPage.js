import React from 'react';
import ExpenseList from './ExpenseList.js';
import ExpenseListFilters from './ExpenseListFilters.js';
import ExpensesSummary  from './ExpensesSummary.js';


const ExpenseDashboardPage = ()=>(
  <div>
    <ExpenseListFilters/>
    <ExpensesSummary/>
    <ExpenseList/>
  </div>
);

export default ExpenseDashboardPage;