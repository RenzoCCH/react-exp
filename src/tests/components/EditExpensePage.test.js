import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense, removeExpense, history, wrapper, match;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  match = {params:{id:expenses[1].id}};
  wrapper = shallow(<EditExpensePage
    params = {expenses[1]}
    expense={expenses[1]}
    editExpense={editExpense}
    removeExpense={removeExpense}
    history={history}
    match={match}
  />);
});

test('Should render EditExpensePage', ()=> {
  expect(wrapper).toMatchSnapshot();
});

test('Should Handle EditExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1]);
});

test('Should Handle removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({
    id: expenses[1].id
  });
});


