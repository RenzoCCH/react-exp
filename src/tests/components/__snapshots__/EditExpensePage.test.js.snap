// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage 1`] = `
<div>
  2
  <ExpenseForm
    expense={
      Object {
        "amount": 109500,
        "createdAt": "1969-12-28T00:00:00.000Z",
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
